File: scripts/build_pose_contact_sheet.py
#!/usr/bin/env python3
import argparse
import json
from pathlib import Path

from PIL import Image, ImageDraw, ImageFont, ImageOps


ROOT = Path(__file__).resolve().parent.parent
DATASET_ROOT = ROOT / "datasets" / "fashion_action_reference"
ANNOTATIONS_DIR = DATASET_ROOT / "pose_annotations"


def read_jsonl(path: Path):
    with path.open("r", encoding="utf-8") as handle:
        return [json.loads(line) for line in handle if line.strip()]


def load_font(size: int):
    candidates = [
        "/System/Library/Fonts/Supplemental/Arial.ttf",
        "/System/Library/Fonts/Helvetica.ttc",
        "/Library/Fonts/Arial.ttf",
    ]
    for candidate in candidates:
        try:
            return ImageFont.truetype(candidate, size)
        except OSError:
            pass
    return ImageFont.load_default()


def make_tile(task, slot: int, tile_w: int, tile_h: int, label_h: int):
    image_area_h = tile_h - label_h
    source = Image.open(task["image"]["local_path"])
    source = ImageOps.exif_transpose(source).convert("RGB")
    source.thumbnail((tile_w, image_area_h), Image.Resampling.LANCZOS)

    tile = Image.new("RGB", (tile_w, tile_h), "white")
    x = (tile_w - source.width) // 2
    y = label_h + (image_area_h - source.height) // 2
    tile.paste(source, (x, y))

    draw = ImageDraw.Draw(tile)
    draw.rectangle((0, 0, tile_w, label_h), fill=(20, 20, 20))
    label = f"{slot} | {task['task_id']} | idx {task['image']['manifest_index']}"
    font = load_font(20)
    draw.text((10, 10), label, fill="white", font=font)
    return tile


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--tasks", default=str(ANNOTATIONS_DIR / "pending_tasks.jsonl"))
    parser.add_argument("--out-dir", default=str(ANNOTATIONS_DIR / "contact_sheets"))
    parser.add_argument("--batch", type=int, default=0)
    parser.add_argument("--count", type=int, default=6)
    parser.add_argument("--columns", type=int, default=3)
    parser.add_argument("--tile-width", type=int, default=430)
    parser.add_argument("--tile-height", type=int, default=620)
    args = parser.parse_args()

    tasks_path = Path(args.tasks)
    out_dir = Path(args.out_dir)
    out_dir.mkdir(parents=True, exist_ok=True)

    tasks = read_jsonl(tasks_path)
    start = args.batch * args.count
    selected = tasks[start : start + args.count]
    if not selected:
        raise SystemExit(f"No tasks selected for batch={args.batch}, count={args.count}")

    label_h = 44
    rows = (len(selected) + args.columns - 1) // args.columns
    sheet = Image.new("RGB", (args.columns * args.tile_width, rows * args.tile_height), "white")
    for index, task in enumerate(selected):
        tile = make_tile(task, index + 1, args.tile_width, args.tile_height, label_h)
        x = (index % args.columns) * args.tile_width
        y = (index // args.columns) * args.tile_height
        sheet.paste(tile, (x, y))

    first_index = selected[0]["image"]["manifest_index"]
    last_index = selected[-1]["image"]["manifest_index"]
    stem = f"batch_{args.batch:05d}_idx_{first_index:05d}_{last_index:05d}"
    sheet_path = out_dir / f"{stem}.png"
    manifest_path = out_dir / f"{stem}.json"
    sheet.save(sheet_path, "PNG")
    manifest = {
        "batch": args.batch,
        "start": start,
        "count": len(selected),
        "columns": args.columns,
        "rows": rows,
        "tile_width": args.tile_width,
        "tile_height": args.tile_height,
        "sheet_path": str(sheet_path.resolve()),
        "tasks": [{"sheet_slot": i + 1, **task} for i, task in enumerate(selected)],
    }
    manifest_path.write_text(json.dumps(manifest, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")
    print(json.dumps({"sheet_path": str(sheet_path.resolve()), "manifest_path": str(manifest_path.resolve()), "count": len(selected)}, ensure_ascii=False, indent=2))


if __name__ == "__main__":
    main()

File: scripts/build_pose_contact_sheet.mjs
import fs from "node:fs/promises";
import path from "node:path";
import { spawn } from "node:child_process";
import { DATASET_ROOT } from "./fashion_tools.mjs";

const args = new Map();
for (let i = 2; i < process.argv.length; i += 2) {
  const key = process.argv[i];
  const value = process.argv[i + 1];
  if (!key?.startsWith("--")) continue;
  args.set(key.slice(2), value ?? "true");
}

const annotationsDir = path.join(DATASET_ROOT, "pose_annotations");
const tasksPath = path.resolve(args.get("tasks") || path.join(annotationsDir, "pending_tasks.jsonl"));
const outDir = path.resolve(args.get("out-dir") || path.join(annotationsDir, "contact_sheets"));
const batch = Math.max(0, Number(args.get("batch") || 0));
const count = Math.max(1, Number(args.get("count") || 6));
const columns = Math.max(1, Number(args.get("columns") || 3));
const tileW = Math.max(240, Number(args.get("tile-width") || 420));
const tileH = Math.max(280, Number(args.get("tile-height") || 560));
const labelH = 44;

async function readJsonl(file) {
  const text = await fs.readFile(file, "utf8");
  return text.split(/\n/).filter(Boolean).map((line) => JSON.parse(line));
}

function run(command, args) {
  return new Promise((resolve, reject) => {
    const child = spawn(command, args, { stdio: ["ignore", "pipe", "pipe"] });
    let stdout = "";
    let stderr = "";
    child.stdout.on("data", (data) => { stdout += data; });
    child.stderr.on("data", (data) => { stderr += data; });
    child.on("close", (code) => {
      if (code === 0) resolve({ stdout, stderr });
      else reject(new Error(`${command} exited ${code}\n${stderr}`));
    });
  });
}

function ffmpegText(value) {
  return String(value)
    .replace(/\\/g, "\\\\")
    .replace(/:/g, "\\:")
    .replace(/'/g, "\\'");
}

const tasks = await readJsonl(tasksPath);
const start = batch * count;
const selected = tasks.slice(start, start + count);
if (!selected.length) {
  console.error(`No tasks selected for batch=${batch}, count=${count}`);
  process.exit(2);
}

await fs.mkdir(outDir, { recursive: true });
const tempDir = path.join(outDir, `.tmp_batch_${String(batch).padStart(5, "0")}`);
await fs.rm(tempDir, { recursive: true, force: true });
await fs.mkdir(tempDir, { recursive: true });

const tilePaths = [];
for (const [i, task] of selected.entries()) {
  const label = `${i + 1} | ${task.task_id} | idx ${task.image.manifest_index}`;
  const out = path.join(tempDir, `tile_${String(i).padStart(2, "0")}.png`);
  const vf = [
    `scale=${tileW}:${tileH - labelH}:force_original_aspect_ratio=decrease`,
    `pad=${tileW}:${tileH - labelH}:(ow-iw)/2:(oh-ih)/2:white`,
    `pad=${tileW}:${tileH}:0:${labelH}:white`,
    `drawbox=x=0:y=0:w=${tileW}:h=${labelH}:color=black@0.82:t=fill`,
    `drawtext=text='${ffmpegText(label)}':x=10:y=10:fontsize=20:fontcolor=white`,
  ].join(",");
  await run("ffmpeg", ["-hide_banner", "-loglevel", "error", "-y", "-i", task.image.local_path, "-vf", vf, out]);
  tilePaths.push(out);
}

const rows = Math.ceil(selected.length / columns);
const inputs = tilePaths.flatMap((file) => ["-i", file]);
const layout = selected.map((_, i) => {
  const x = (i % columns) * tileW;
  const y = Math.floor(i / columns) * tileH;
  return `${x}_${y}`;
}).join("|");
const sheetPath = path.join(outDir, `batch_${String(batch).padStart(5, "0")}.png`);
await run("ffmpeg", [
  "-hide_banner",
  "-loglevel",
  "error",
  "-y",
  ...inputs,
  "-filter_complex",
  `xstack=inputs=${selected.length}:layout=${layout}:fill=white`,
  sheetPath,
]);

const manifestPath = path.join(outDir, `batch_${String(batch).padStart(5, "0")}.json`);
await fs.writeFile(manifestPath, `${JSON.stringify({
  batch,
  start,
  count: selected.length,
  columns,
  rows,
  tile_width: tileW,
  tile_height: tileH,
  sheet_path: sheetPath,
  tasks: selected.map((task, i) => ({ sheet_slot: i + 1, ...task })),
}, null, 2)}\n`, "utf8");
await fs.rm(tempDir, { recursive: true, force: true });

console.log(JSON.stringify({ batch, start, count: selected.length, sheet_path: sheetPath, manifest_path: manifestPath }, null, 2));
